modules/502-delivery/hooks/werf_sources/hook.go

```diff
@@ -0,0 +1,406 @@
+	if err != nil {
+		return nil, err
+	}
+	if !ok {
+		return nil, fmt.Errorf("spec.imageRepo field expected")
+	}
+
+	ws.apiURL, ok, err = unstructured.NestedString(obj.Object, "spec", "apiUrl")
+	if err != nil {
+		return nil, err
+	}
+	if !ok {
+		ws.apiURL = "https://" + firstSegment(ws.repo)
+	}
+
+	ws.pullSecretName, _, err = unstructured.NestedString(obj.Object, "spec", "pullSecretName")
+	if err != nil {
+		return nil, err
+	}
+
+	// By default, Argo CD is desired, but the OCI repo can be disabled to use purely Image
+	// Updater functionality along with another repository type.
+	repoEnabled, ok, err := unstructured.NestedBool(obj.Object, "spec", "argocdRepoEnabled")
+	if err != nil {
+		return nil, err
+	}
+	if !ok {
+		repoEnabled = true
+	}
+
+	// By default, Argo CD repo belongs to the "default" project.
+	arepo, ok, err := unstructured.NestedStringMap(obj.Object, "spec", "argocdRepo")
+	if err != nil {
+		return nil, err
+	}
+	project := "default"
+	if repoEnabled && ok {
+		specifiedProject, projectSpecified := arepo["project"]
+		if projectSpecified && specifiedProject != "" {
+			project = specifiedProject
+		}
+	}
+	if repoEnabled {
+		ws.argocdRepo = &argocdRepoConfig{
+			project: project,
+		}
+	}
+
+	return ws, nil
+}
+
+func parseWerfSources(snapshots []go_hook.FilterResult) ([]werfSource, error) {
+	var res []werfSource
+	for _, snap := range snapshots {
+		r, ok := snap.(werfSource)
+		if !ok {
+			return nil, fmt.Errorf("unexpected type %T", snap)
+		}
+		res = append(res, r)
+	}
+	return res, nil
+}
+
+type registryCredentials struct {
+	username string
+	password string
+}
+
+func parseDockerConfigJSONCredentials(config dockerFileConfig, registry string) (registryCredentials, error) {
+	creds := registryCredentials{}
+
+	cfg, ok := config.Auths[registry]
+	if !ok {
+		return creds, fmt.Errorf("no credentials")
+	}
+
+	if cfg.Auth != "" {
+		auth, err := base64.StdEncoding.DecodeString(cfg.Auth)
+		if err != nil {
+			return creds, fmt.Errorf(`cannot decode base64 "auth" field`)
+		}
+		parts := strings.Split(string(auth), ":")
+		if len(parts) != 2 {
+			return creds, fmt.Errorf(`unexpected format of "auth" field, expected "username:password"`)
+		}
+		creds.username, creds.password = parts[0], parts[1]
+		return creds, nil
+	}
+
+	creds.username, creds.password = cfg.Username, cfg.Password
+	return creds, nil
+}
+
+/*
+	{ "auths":{
+	        "cr.example.com":{
+			"username":"...",
+			"password":"...",
+			"auth":"base64([username]:[password])",
+			"email":"...@example.com"
+		}
+	}}
+*/
+type dockerFileConfig struct {
+	Auths map[string]authn.AuthConfig `json:"auths"`
+}
+
+type credSecret struct {
+	Name   string
+	Config dockerFileConfig
+}
+
+func filterDockerConfigJSON(obj *unstructured.Unstructured) (go_hook.FilterResult, error) {
+	var secret corev1.Secret
+	err := sdk.FromUnstructured(obj, &secret)
+	if err != nil {
+		return nil, err
+	}
+
+	if secret.Type != corev1.SecretTypeDockerConfigJson {
```

Comment:
NIT: Type can be a field selector
```
kubectl get secret --field-selector=type='kubernetes.io/dockerconfigjson' -A
```

Reply:
Let's leave a comment at least that it is required because of the way tests are working.

Reply:
did it